# Project 2020-2021 - *Foundations of Computer Science*

### Group 98 - *Irene Giannuzzi, Federica Zaccagnini* 

You have to work on the [University dataset](https://drive.google.com/drive/folders/1Hs3nRtK_F3h8eg59B4-TD1DEua6g8Klv?usp=sharing). It contains three different university rankings:
*  The Times Higher Education World University Ranking, shortly *Times*,
*  the Academic Ranking of World Universities, shortly *Shanghai*,
*  the Center for World University Rankings, shortly *cwur*.

### Notes

1.    It is mandatory to use GitHub for developing the project.
1.    The project must be a jupyter notebook.
1.    There is no restriction on the libraries that can be used, nor on the Python version.
1.    All questions on the project **must** be asked in a public channel on [Zulip](https://focs.zulipchat.com).

In [1]:
# pip install unidecode

In [2]:
import pandas as pd
import numpy as np
import unidecode
import re

## Caricamento dati 

In [3]:
times = pd.read_csv('timesData.csv', na_values = '-', thousands = ',')
times

world_rank                        university_name  \
0             1                     Harvard University   
1             2     California Institute of Technology   
2             3  Massachusetts Institute of Technology   
3             4                    Stanford University   
4             5                   Princeton University   
...         ...                                    ...   
2598    601-800                    Yeungnam University   
2599    601-800            Yıldız Technical University   
2600    601-800               Yokohama City University   
2601    601-800           Yokohama National University   
2602    601-800                     Yuan Ze University   

                       country  teaching  international  research  citations  \
0     United States of America      99.7           72.4      98.7       98.8   
1     United States of America      97.7           54.6      98.0       99.9   
2     United States of America      97.8           82.3      91.4       99.9   
3     United States of America      98.3           29.5      98.1       99.2   
4     United States of America      90.9           70.3      95.4       99.9   
...                        ...       ...            ...       ...        ...   
2598               South Korea      18.6           24.3      10.9       26.5   
2599                    Turkey      14.5           14.9       7.6       19.3   
2600                     Japan      24.0           16.1      10.2       36.4   
2601                     Japan      20.1           23.3      16.0       13.5   
2602                    Taiwan      16.2           17.7      18.3       28.6   

      income  total_score  num_students  student_staff_ratio  \
0       34.5         96.1       20152.0                  8.9   
1       83.7         96.0        2243.0                  6.9   
2       87.5         95.6       11074.0                  9.0   
3       64.3         94.3       15596.0                  7.8   
4        NaN         94.2        7929.0                  8.4   
...      ...          ...           ...                  ...   
2598    35.4          NaN       21958.0                 15.3   
2599    44.0          NaN       31268.0                 28.7   
2600    37.9          NaN        4122.0                  3.7   
2601    40.4          NaN       10117.0                 12.1   
2602    39.8          NaN        8663.0                 20.6   

     international_students female_male_ratio  year  
0                       25%               NaN  2011  
1                       27%           33 : 67  2011  
2                       33%           37 : 63  2011  
3                       22%           42 : 58  2011  
4                       27%           45 : 55  2011  
...                     ...               ...   ...  
2598                     3%           48 : 52  2016  
2599                     2%           36 : 64  2016  
2600                     3%               NaN  2016  
2601                     8%           28 : 72  2016  
2602                     4%           43 : 57  2016  

[2603 rows x 14 columns]

In [4]:
shanghai = pd.read_csv('shanghaiData.csv')
shanghai

world_rank                              university_name national_rank  \
0             1                           Harvard University             1   
1             2                      University of Cambridge             1   
2             3                          Stanford University             2   
3             4           University of California, Berkeley             3   
4             5  Massachusetts Institute of Technology (MIT)             4   
...         ...                                          ...           ...   
4892    401-500                        University of Trieste         11-20   
4893    401-500                       University of Zaragoza          9-13   
4894    401-500                        Utah State University       126-146   
4895    401-500              Vienna University of Technology           4-6   
4896    401-500                       Wake Forest University       126-146   

      total_score  alumni  award   hici     ns    pub   pcp  year  
0           100.0   100.0  100.0  100.0  100.0  100.0  72.4  2005  
1            73.6    99.8   93.4   53.3   56.6   70.9  66.9  2005  
2            73.4    41.1   72.2   88.5   70.9   72.3  65.0  2005  
3            72.8    71.8   76.0   69.4   73.9   72.2  52.7  2005  
4            70.1    74.0   80.6   66.7   65.8   64.3  53.0  2005  
...           ...     ...    ...    ...    ...    ...   ...   ...  
4892          NaN     0.0    0.0    5.0   10.9   25.1  20.1  2015  
4893          NaN     0.0    0.0    7.6    5.1   33.3  13.1  2015  
4894          NaN    13.6    0.0    3.6   10.8   25.1  15.5  2015  
4895          NaN     0.0    0.0    0.0   12.2   28.8  22.9  2015  
4896          NaN     0.0    0.0   14.9    7.5   25.0  11.9  2015  

[4897 rows x 11 columns]

In [5]:
cwur = pd.read_csv('cwurData.csv')
cwur = cwur.rename(columns = {'institution':'university_name'}) # rinominiamo colonna con i nomi delle università per merge successivi
cwur

world_rank                        university_name         country  \
0              1                     Harvard University             USA   
1              2  Massachusetts Institute of Technology             USA   
2              3                    Stanford University             USA   
3              4                University of Cambridge  United Kingdom   
4              5     California Institute of Technology             USA   
...          ...                                    ...             ...   
2195         996              University of the Algarve        Portugal   
2196         997                  Alexandria University           Egypt   
2197         998            Federal University of Ceará          Brazil   
2198         999                 University of A Coruña           Spain   
2199        1000        China Pharmaceutical University           China   

      national_rank  quality_of_education  alumni_employment  \
0                 1                     7                  9   
1                 2                     9                 17   
2                 3                    17                 11   
3                 1                    10                 24   
4                 4                     2                 29   
...             ...                   ...                ...   
2195              7                   367                567   
2196              4                   236                566   
2197             18                   367                549   
2198             40                   367                567   
2199             83                   367                567   

      quality_of_faculty  publications  influence  citations  broad_impact  \
0                      1             1          1          1           NaN   
1                      3            12          4          4           NaN   
2                      5             4          2          2           NaN   
3                      4            16         16         11           NaN   
4                      7            37         22         22           NaN   
...                  ...           ...        ...        ...           ...   
2195                 218           926        845        812         969.0   
2196                 218           997        908        645         981.0   
2197                 218           830        823        812         975.0   
2198                 218           886        974        812         975.0   
2199                 218           861        991        812         981.0   

      patents   score  year  
0           5  100.00  2012  
1           1   91.67  2012  
2          15   89.50  2012  
3          50   86.17  2012  
4          18   85.21  2012  
...       ...     ...   ...  
2195      816   44.03  2015  
2196      871   44.03  2015  
2197      824   44.03  2015  
2198      651   44.02  2015  
2199      547   44.02  2015  

[2200 rows x 14 columns]

Importiamo anche la tabella con i nomi delle università e il paese a cui fanno riferimento per verificare la correttezza dei nomi delle università riportati nei tre dataset.

In [6]:
school_country_table = pd.read_csv('school_and_country_table.csv')
school_country_table

school_name                   country
0                       Harvard University  United States of America
1       California Institute of Technology  United States of America
2    Massachusetts Institute of Technology  United States of America
3                      Stanford University  United States of America
4                     Princeton University  United States of America
..                                     ...                       ...
813                      Xidian University                     China
814                    Yeungnam University               South Korea
815            Yıldız Technical University                    Turkey
816               Yokohama City University                     Japan
817           Yokohama National University                     Japan

[818 rows x 2 columns]

## Pre-processing

Qui eseguiamo tutte le operazioni di aggiustamento dati utili per lo svolgimento dei punti successivi. 

Innanzitutto trasformiamo la colonna `income` di `times` in numeric dato che ci servirà nel punto 2. 

In [7]:
times['income'] = pd.to_numeric(times['income'])

Convertiamo in stringhe tutti i nomi delle università.

In [8]:
times['university_name'] = times['university_name'].astype(str)
shanghai['university_name'] = shanghai['university_name'].astype(str)
cwur['university_name'] = cwur['university_name'].astype(str)

Riteniamo ragionevole rendere uniformi i nomi delle università in quanto, se analizziamo attentamente i tre dataset a disposizione e li confrontiamo con il DataFrame `school_country_table`, notiamo delle incongruenze. Costruiamo 4 liste inerenti ai nomi delle università per ciascun dataset e proviamo a confrontarle. 

In [9]:
# Utilizziamo set per non avere duplicati
universities_list = set(list(school_country_table['school_name']))
times_universities = set(list(times['university_name']))
shanghai_universities = set(list(shanghai['university_name']))
cwur_universities = set(list(cwur['university_name']))

Contiamo quanti nomi corrispondono: 

In [10]:
count = 0
for university in universities_list:
    if university in times_universities:
        count+=1
print(count)

818


In [11]:
count = 0
for university in universities_list:
    if university in shanghai_universities:
        count+=1
print(count)

384


In [12]:
count = 0
for university in universities_list:
    if university in cwur_universities:
        count+=1
print(count)

589


Nel DataFrame `times`, tutti i nomi delle università corrispondono a quelli in `school_country_table`. 
Nel DataFrame `shanghai`, solo 384 nomi delle università corrispondono a quelli in `school_country_table`. 
Nel DataFrame `cwur`, solo 589 nomi delle università corrispondono a quelli in `school_country_table`. 

Ora contiamo quante università abbiamo nei vari dataset: 

In [13]:
len(times_universities)

818

In [14]:
len(shanghai_universities)

659

In [15]:
len(cwur_universities)

1024

Arriviamo alle seguenti conclusioni:

* Alcune università che sono presenti in `school_country_table` e `times` potrebbero non essere presenti nei dataframe `shanghai` e `cwur`. 

* Nulla garantisce che le 818 università in `times` siano esattamente le stesse di quelle contenute in `school_country_table`. 

* Alcune università potrebbero essere nominate in modo differente nei dataframe `shanghai` e `cwur` rispetto ai nomi presenti in `school_country_table` e `times`.

* Alcune università, nello stesso DataFrame, potrebbero essere nominate in modo diverso nonostante esse rappresentino la stessa università.

Al fine di combinare correttamente i tre dataset a disposizione per lo svoglimento dei punti successivi, decidiamo di uniformare i nomi delle università svolgendo le seguenti operazioni:

1. Rimozione punteggiatura e numeri e sostituzione con un singolo spazio. 
2. Rimozione spazi iniziali e finali.
3. Rimozione doppi spazi (o più) e sostituzione con un singolo spazio.
4. Trasformazione caratteri in minuscolo per non avere problemi di duplicati.
5. Rimozione caratteri accentati per uniformare il più possibile i nomi delle università.

In [16]:
times['university_name'] = times['university_name'].apply(lambda row: re.sub('[^a-zA-Z]', ' ', row)) # tutto ciò che non è lettere
times['university_name'] = times['university_name'].str.strip()                                      # spazi iniziali e finali
times['university_name'] = times['university_name'].apply(lambda row: re.sub('\s+', ' ', row))       # spazi multipli
times['university_name'] = times['university_name'].str.lower()                                      # minuscolo
times["university_name"] = times['university_name'].apply(unidecode.unidecode)                       # caratteri accentati

In [17]:
shanghai['university_name'] = shanghai['university_name'].apply(lambda row: re.sub('[^a-zA-Z]', ' ', row))    # tutto ciò che non è lettere
shanghai['university_name'] = shanghai['university_name'].str.strip()                                         # spazi iniziali e finali
shanghai['university_name'] = shanghai['university_name'].apply(lambda row: re.sub('\s+', ' ', row))          # spazi multipli
shanghai['university_name'] = shanghai['university_name'].str.lower()                                         # minuscolo
shanghai["university_name"] = shanghai['university_name'].apply(unidecode.unidecode)                          # caratteri accentati

In [18]:
cwur['university_name'] = cwur['university_name'].apply(lambda row: re.sub('[^a-zA-Z]', ' ', row)) # tutto ciò che non è lettere
cwur['university_name'] = cwur['university_name'].str.strip()                                      # spazi iniziali e finali
cwur['university_name'] = cwur['university_name'].apply(lambda row: re.sub('\s+', ' ', row))       # spazi multipli
cwur['university_name'] = cwur['university_name'].str.lower()                                      # minuscolo
cwur["university_name"] = cwur['university_name'].apply(unidecode.unidecode)                       # caratteri accentati

Svolgendo i punti successivi (in particolare il punto 5), si è notata la presenza di due errori di battitura nel dataset `times`:

* In alcuni casi, è stato scritto *Unisted States of America* anziché *United States of America*.
* In alcuni casi, è stato scritto *Unted Kingdom* anziché *United Kingdom*.

Per questo motivo, correggiamo:

In [19]:
times['country'] = times['country'].str.replace('Unisted States of America', 'United States of America')
times['country'] = times['country'].str.replace('Unted Kingdom', 'United Kingdom')

Controlliamo ora di avere tutti i paesi e che essi siano effettivamente corretti:

In [20]:
len(set(list(school_country_table['country'])))

70

In [21]:
len(set(list(times['country'])))

70

In [22]:
set(list(school_country_table['country']))

{'Argentina',
 'Australia',
 'Austria',
 'Bangladesh',
 'Belarus',
 'Belgium',
 'Brazil',
 'Canada',
 'Chile',
 'China',
 'Colombia',
 'Cyprus',
 'Czech Republic',
 'Denmark',
 'Egypt',
 'Estonia',
 'Finland',
 'France',
 'Germany',
 'Ghana',
 'Greece',
 'Hong Kong',
 'Hungary',
 'Iceland',
 'India',
 'Indonesia',
 'Iran',
 'Israel',
 'Italy',
 'Japan',
 'Jordan',
 'Kenya',
 'Latvia',
 'Lebanon',
 'Lithuania',
 'Luxembourg',
 'Macau',
 'Malaysia',
 'Mexico',
 'Morocco',
 'Netherlands',
 'New Zealand',
 'Nigeria',
 'Norway',
 'Oman',
 'Pakistan',
 'Poland',
 'Portugal',
 'Qatar',
 'Republic of Ireland',
 'Romania',
 'Russian Federation',
 'Saudi Arabia',
 'Serbia',
 'Singapore',
 'Slovakia',
 'Slovenia',
 'South Africa',
 'South Korea',
 'Spain',
 'Sweden',
 'Switzerland',
 'Taiwan',
 'Thailand',
 'Turkey',
 'Uganda',
 'Ukraine',
 'United Arab Emirates',
 'United Kingdom',
 'United States of America'}

In [23]:
set(list(times['country']))

{'Argentina',
 'Australia',
 'Austria',
 'Bangladesh',
 'Belarus',
 'Belgium',
 'Brazil',
 'Canada',
 'Chile',
 'China',
 'Colombia',
 'Cyprus',
 'Czech Republic',
 'Denmark',
 'Egypt',
 'Estonia',
 'Finland',
 'France',
 'Germany',
 'Ghana',
 'Greece',
 'Hong Kong',
 'Hungary',
 'Iceland',
 'India',
 'Indonesia',
 'Iran',
 'Israel',
 'Italy',
 'Japan',
 'Jordan',
 'Kenya',
 'Latvia',
 'Lebanon',
 'Lithuania',
 'Luxembourg',
 'Macau',
 'Malaysia',
 'Mexico',
 'Morocco',
 'Netherlands',
 'New Zealand',
 'Nigeria',
 'Norway',
 'Oman',
 'Pakistan',
 'Poland',
 'Portugal',
 'Qatar',
 'Republic of Ireland',
 'Romania',
 'Russian Federation',
 'Saudi Arabia',
 'Serbia',
 'Singapore',
 'Slovakia',
 'Slovenia',
 'South Africa',
 'South Korea',
 'Spain',
 'Sweden',
 'Switzerland',
 'Taiwan',
 'Thailand',
 'Turkey',
 'Uganda',
 'Ukraine',
 'United Arab Emirates',
 'United Kingdom',
 'United States of America'}

### 1. For each university, extract from the *times* dataset the most recent and the least recent data, obtaining two separate dataframes

Considerando la colonna `year` e la colonna `university_name`, estraiamo due DataFrames:
* `recent`, contenente i dati più recenti
* `old`, contenente i dati meno recenti

In [24]:
recent = times.loc[times.groupby('university_name')['year'].idxmax()]
recent

world_rank               university_name         country  teaching  \
2003    201-250            aalborg university         Denmark      25.1   
2056    251-300              aalto university         Finland      31.1   
1908       =106             aarhus university         Denmark      36.9   
2105    301-350        aberystwyth university  United Kingdom      21.6   
2404    601-800    adam mickiewicz university          Poland      20.0   
...         ...                           ...             ...       ...   
2601    601-800  yokohama national university           Japan      20.1   
2151    301-350             yonsei university     South Korea      37.2   
2152    301-350               york university          Canada      30.2   
2602    601-800            yuan ze university          Taiwan      16.2   
2104    251-300           zhejiang university           China      44.3   

      international  research  citations  income  total_score  num_students  \
2003           71.0      28.4       73.8    43.7          NaN       17422.0   
2056           65.4      32.8       62.1    61.6          NaN       16099.0   
1908           76.8      50.7       79.8    68.3         57.7       23895.0   
2105           72.2      18.9       67.2    31.3          NaN        9252.0   
2404           25.7      11.0       15.3    28.7          NaN       40633.0   
...             ...       ...        ...     ...          ...           ...   
2601           23.3      16.0       13.5    40.4          NaN       10117.0   
2151           41.5      34.1       39.4    75.9          NaN       24774.0   
2152           58.2      30.8       49.1    33.2          NaN       48007.0   
2602           17.7      18.3       28.6    39.8          NaN        8663.0   
2104           19.6      46.0       36.1    96.2          NaN       47508.0   

      student_staff_ratio international_students female_male_ratio  year  
2003                 15.9                    15%           48 : 52  2016  
2056                 24.2                    17%           32 : 68  2016  
1908                 13.6                    14%           54 : 46  2016  
2105                 19.2                    18%           48 : 52  2016  
2404                 15.6                     1%           71 : 29  2016  
...                   ...                    ...               ...   ...  
2601                 12.1                     8%           28 : 72  2016  
2151                 11.6                    14%               NaN  2016  
2152                 39.4                     9%           59 : 41  2016  
2602                 20.6                     4%           43 : 57  2016  
2104                 15.9                     5%           41 : 59  2016  

[818 rows x 14 columns]

In [25]:
old = times.loc[times.groupby('university_name')['year'].idxmin()]
old

world_rank               university_name         country  teaching  \
501     301-350            aalborg university         Denmark      19.0   
502     301-350              aalto university         Finland      26.2   
166         167             aarhus university         Denmark      38.1   
476     276-300        aberystwyth university  United Kingdom      19.8   
2404    601-800    adam mickiewicz university          Poland      20.0   
...         ...                           ...             ...       ...   
2601    601-800  yokohama national university           Japan      20.1   
191         190             yonsei university     South Korea      43.0   
500     276-300               york university          Canada      19.9   
601     350-400            yuan ze university          Taiwan      10.8   
197         197           zhejiang university           China      54.6   

      international  research  citations  income  total_score  num_students  \
501            75.3      20.0       27.1    36.4          NaN       17422.0   
502            49.0      22.2       37.5    61.9          NaN       16099.0   
166            33.4      55.6       57.3    61.5         49.9       23895.0   
476            63.8      15.5       56.6    35.5          NaN        9252.0   
2404           25.7      11.0       15.3    28.7          NaN       40633.0   
...             ...       ...        ...     ...          ...           ...   
2601           23.3      16.0       13.5    40.4          NaN       10117.0   
191            28.0      48.7       52.2    40.4         46.9       24774.0   
500            57.7      27.7       41.2    41.7          NaN       48007.0   
601            12.8       9.6       58.3    29.2          NaN        8663.0   
197            29.6      41.3       44.3    70.3         46.4       47508.0   

      student_staff_ratio international_students female_male_ratio  year  
501                  15.9                    15%           48 : 52  2012  
502                  24.2                    17%           32 : 68  2012  
166                  13.6                    14%           54 : 46  2011  
476                  19.2                    18%           48 : 52  2012  
2404                 15.6                     1%           71 : 29  2016  
...                   ...                    ...               ...   ...  
2601                 12.1                     8%           28 : 72  2016  
191                  11.6                    14%               NaN  2011  
500                  39.4                     9%           59 : 41  2012  
601                  20.6                     4%           43 : 57  2012  
197                  15.9                     5%           41 : 59  2011  

[818 rows x 14 columns]

### 2. For each university, compute the improvement in `income` between the least recent and the most recent data points

Per calcolare il guadagno, occorre unire i due DataFrames utilizzando la colonna `university_name` ed effettuare semplicemente una differenza tra le colonne `income_recent` e `income_old`.

In [26]:
old_and_recent = pd.merge(old, recent, on = 'university_name', suffixes = ['_old', '_recent'])
old_and_recent[['university_name', 'income_old', 'income_recent']]

university_name  income_old  income_recent
0              aalborg university        36.4           43.7
1                aalto university        61.9           61.6
2               aarhus university        61.5           68.3
3          aberystwyth university        35.5           31.3
4      adam mickiewicz university        28.7           28.7
..                            ...         ...            ...
813  yokohama national university        40.4           40.4
814             yonsei university        40.4           75.9
815               york university        41.7           33.2
816            yuan ze university        29.2           39.8
817           zhejiang university        70.3           96.2

[818 rows x 3 columns]

In [27]:
old_and_recent['improvement'] = old_and_recent['income_recent'] - old_and_recent['income_old']
old_and_recent[['university_name', 'income_old', 'income_recent', 'improvement']]

university_name  income_old  income_recent  improvement
0              aalborg university        36.4           43.7          7.3
1                aalto university        61.9           61.6         -0.3
2               aarhus university        61.5           68.3          6.8
3          aberystwyth university        35.5           31.3         -4.2
4      adam mickiewicz university        28.7           28.7          0.0
..                            ...         ...            ...          ...
813  yokohama national university        40.4           40.4          0.0
814             yonsei university        40.4           75.9         35.5
815               york university        41.7           33.2         -8.5
816            yuan ze university        29.2           39.8         10.6
817           zhejiang university        70.3           96.2         25.9

[818 rows x 4 columns]

### 3. Find the university with the largest increase computed in the previous point

In [28]:
old_and_recent.loc[old_and_recent['improvement'].idxmax()]['university_name']

'tu dresden'

Controlliamo se c'è più di una università con lo stesso guadagno massimo:

In [29]:
old_and_recent[old_and_recent['improvement'] == old_and_recent['improvement'].max()][['university_name', 'improvement']]

university_name  improvement
470      tu dresden         67.8

Non sono presenti altre università che hanno guadagno pari a quello massimo. Pertanto, anche solo la prima soluzione va bene.

### 4. For each ranking, consider only the most recent data point. For each university, compute the maximum difference between the rankings (e.g. for *Aarhus University* the value is 122-73=49). Notice that some rankings are expressed as a range

Innanzitutto estraiamo i dati più recenti anche dai dataset `shangai` e `cwur`. 

In [30]:
recent_shanghai = shanghai.loc[shanghai.groupby('university_name')['year'].idxmax()]
recent_shanghai

world_rank                       university_name national_rank  \
4697    301-400                    aalborg university             5   
4797    401-500                      aalto university           4-6   
4469         73                     aarhus university             2   
4497    101-150              aix marseille university           5-6   
4798    401-500  aristotle university of thessaloniki             2   
...         ...                                   ...           ...   
3512    401-500                  yamaguchi university         17-23   
4695    201-300                    yeshiva university        79-102   
4696    201-300                     yonsei university           2-5   
4396    401-500                       york university         19-21   
4546    101-150                   zhejiang university           1-4   

      total_score  alumni  award  hici    ns   pub   pcp  year  
4697          NaN     0.0    0.0  11.2   4.6  30.4  16.8  2015  
4797          NaN     0.0    0.0   0.0  10.4  33.4  17.0  2015  
4469         27.3    11.5   22.1  12.3  25.8  51.8  31.0  2015  
4497          NaN    13.6    0.0  15.2  20.3  50.1  27.6  2015  
4798          NaN     0.0    0.0   4.9   1.5  34.0  16.8  2015  
...           ...     ...    ...   ...   ...   ...   ...   ...  
3512          NaN     0.0    0.0  16.1   2.8  20.0  13.2  2011  
4695          NaN     0.0    0.0   6.3  21.7  31.5  21.6  2015  
4696          NaN     0.0    0.0   8.6   4.9  52.1  21.4  2015  
4396          NaN     0.0    0.0   0.0  10.1  30.6  12.4  2014  
4546          NaN     0.0    0.0   7.1  13.5  68.5  19.9  2015  

[650 rows x 11 columns]

In [31]:
recent_cwur = cwur.loc[cwur.groupby('university_name')['year'].idxmax()]
recent_cwur

world_rank                      university_name         country  \
1764         565                   aalborg university         Denmark   
1620         421                     aalto university         Finland   
1321         122                    aarhus university         Denmark   
2013         814               aberystwyth university  United Kingdom   
2089         890  adam mickiewicz university in pozna          Poland   
...          ...                                  ...             ...   
2178         979                   yuan ze university          Taiwan   
1190         991           zhejiang normal university           China   
1390         191                  zhejiang university           China   
2183         984    zhejiang university of technology           China   
1994         795                 zhengzhou university           China   

      national_rank  quality_of_education  alumni_employment  \
1764              5                   367                439   
1620              4                   367                224   
1321              2                   310                165   
2013             59                   367                567   
2089              5                   150                567   
...             ...                   ...                ...   
2178             20                   367                567   
1190             81                   355                478   
1390              4                   367                309   
2183             79                   367                567   
1994             47                   367                427   

      quality_of_faculty  publications  influence  citations  broad_impact  \
1764                 218           596        660        287         520.0   
1620                 218           336        403        645         424.0   
1321                 106            91         81        101          95.0   
2013                 218           842        552        511         754.0   
2089                 218           675        775        812         932.0   
...                  ...           ...        ...        ...           ...   
2178                 218           923        984        812         946.0   
1190                 210           905        932        800         979.0   
1390                 218            71        290        368         265.0   
2183                 218           858        991        812         958.0   
1994                 218           663        806        812         754.0   

      patents  score  year  
1764      401  44.59  2015  
1620      246  45.10  2015  
1321      154  49.65  2015  
2013      607  44.22  2015  
2089      657  44.14  2015  
...       ...    ...   ...  
2178      871  44.05  2015  
1190      737  44.26  2014  
1390      106  47.66  2015  
2183      672  44.04  2015  
1994      749  44.25  2015  

[1022 rows x 14 columns]

Consideriamo anche i dati più recenti estratti da `times` al punto 1.

In [32]:
recent

world_rank               university_name         country  teaching  \
2003    201-250            aalborg university         Denmark      25.1   
2056    251-300              aalto university         Finland      31.1   
1908       =106             aarhus university         Denmark      36.9   
2105    301-350        aberystwyth university  United Kingdom      21.6   
2404    601-800    adam mickiewicz university          Poland      20.0   
...         ...                           ...             ...       ...   
2601    601-800  yokohama national university           Japan      20.1   
2151    301-350             yonsei university     South Korea      37.2   
2152    301-350               york university          Canada      30.2   
2602    601-800            yuan ze university          Taiwan      16.2   
2104    251-300           zhejiang university           China      44.3   

      international  research  citations  income  total_score  num_students  \
2003           71.0      28.4       73.8    43.7          NaN       17422.0   
2056           65.4      32.8       62.1    61.6          NaN       16099.0   
1908           76.8      50.7       79.8    68.3         57.7       23895.0   
2105           72.2      18.9       67.2    31.3          NaN        9252.0   
2404           25.7      11.0       15.3    28.7          NaN       40633.0   
...             ...       ...        ...     ...          ...           ...   
2601           23.3      16.0       13.5    40.4          NaN       10117.0   
2151           41.5      34.1       39.4    75.9          NaN       24774.0   
2152           58.2      30.8       49.1    33.2          NaN       48007.0   
2602           17.7      18.3       28.6    39.8          NaN        8663.0   
2104           19.6      46.0       36.1    96.2          NaN       47508.0   

      student_staff_ratio international_students female_male_ratio  year  
2003                 15.9                    15%           48 : 52  2016  
2056                 24.2                    17%           32 : 68  2016  
1908                 13.6                    14%           54 : 46  2016  
2105                 19.2                    18%           48 : 52  2016  
2404                 15.6                     1%           71 : 29  2016  
...                   ...                    ...               ...   ...  
2601                 12.1                     8%           28 : 72  2016  
2151                 11.6                    14%               NaN  2016  
2152                 39.4                     9%           59 : 41  2016  
2602                 20.6                     4%           43 : 57  2016  
2104                 15.9                     5%           41 : 59  2016  

[818 rows x 14 columns]

Per ogni università, dobbiamo calcolare la differenza massima tra i rankings. Facciamo due considerazioni in merito alla colonna `world_rank`:

* Alcuni valori sono espressi come *range*.
* Alcuni valori hanno un = davanti, che sta ad indicare un pari merito.

Decidiamo di gestire queste due situazioni nel modo seguente:

* Laddove il ranking venga espresso in termini di *range*, consideriamo una semplice media tra il valore massimo e il valore minimo. 
* Laddove ci sia un pari merito, quindi un =, decidiamo di tenere in considerazione semplicemente il valore numerico (ad es., se abbiamo =106 consideriamo solo 106). 

Al fine di effettuare queste operazioni costruiamo la funzione `adjust_rankings`. 

In [33]:
def adjust_rankings(x):
    if '=' in x:
        return x.replace('=','')
    else:
        if '-' in x:
            values = x.split('-')
            return (int(values[0])+int(values[1]))/2
        else:
            return x

In [34]:
recent['world_rank'] = recent['world_rank'].astype(str)
recent_shanghai['world_rank'] = recent_shanghai['world_rank'].astype(str)
recent_cwur['world_rank'] = recent_cwur['world_rank'].astype(str)

In [35]:
recent['world_rank'] = recent['world_rank'].apply(adjust_rankings)
recent_shanghai['world_rank'] = recent_shanghai['world_rank'].apply(adjust_rankings)
recent_cwur['world_rank'] = recent_cwur['world_rank'].apply(adjust_rankings)

Ora dobbiamo unire i DataFrame utilizzando i nomi delle università.

In [36]:
rankings = pd.merge(recent[['university_name', 'world_rank']], recent_shanghai[['university_name', 'world_rank']], on = 'university_name')
rankings.rename(columns = {'world_rank_x':'world_rank_times', 'world_rank_y':'world_rank_shanghai'}, inplace = True)
rankings

university_name world_rank_times world_rank_shanghai
0                      aalborg university            225.5               350.5
1                        aalto university            275.5               450.5
2                       aarhus university              106                  73
3                aix marseille university            275.5               125.5
4    aristotle university of thessaloniki            700.5               450.5
..                                    ...              ...                 ...
394                       yale university               12                  11
395                    yeshiva university              164               250.5
396                     yonsei university            325.5               250.5
397                       york university            325.5               450.5
398                   zhejiang university            275.5               125.5

[399 rows x 3 columns]

In [37]:
rankings = pd.merge(rankings, recent_cwur[['university_name', 'world_rank']], on = 'university_name')
rankings.rename(columns = {'world_rank':'world_rank_cwur'}, inplace = True)
rankings

university_name world_rank_times  \
0                      aalborg university            225.5   
1                        aalto university            275.5   
2                       aarhus university              106   
3                aix marseille university            275.5   
4    aristotle university of thessaloniki            700.5   
..                                    ...              ...   
374                       yale university               12   
375                    yeshiva university              164   
376                     yonsei university            325.5   
377                       york university            325.5   
378                   zhejiang university            275.5   

    world_rank_shanghai world_rank_cwur  
0                 350.5             565  
1                 450.5             421  
2                    73             122  
3                 125.5             206  
4                 450.5             459  
..                  ...             ...  
374                  11              11  
375               250.5             171  
376               250.5              98  
377               450.5             337  
378               125.5             191  

[379 rows x 4 columns]

Ora dobbiamo considerare tutte le coppie di rankings, calcolare le differenze in valore assoluto e tenere quella massima. Per farlo, decidiamo di costruire la funzione `diff(a,b,c)`. 

In [38]:
def diff(a,b,c):
    return max([abs(a-b),abs(a-c),abs(b-c)])

In [39]:
rankings['max_difference'] = rankings.apply(lambda row: diff(pd.to_numeric(row.world_rank_times), pd.to_numeric(row.world_rank_shanghai), pd.to_numeric(row.world_rank_cwur)), axis = 1)
rankings

university_name world_rank_times  \
0                      aalborg university            225.5   
1                        aalto university            275.5   
2                       aarhus university              106   
3                aix marseille university            275.5   
4    aristotle university of thessaloniki            700.5   
..                                    ...              ...   
374                       yale university               12   
375                    yeshiva university              164   
376                     yonsei university            325.5   
377                       york university            325.5   
378                   zhejiang university            275.5   

    world_rank_shanghai world_rank_cwur  max_difference  
0                 350.5             565           339.5  
1                 450.5             421           175.0  
2                    73             122            49.0  
3                 125.5             206           150.0  
4                 450.5             459           250.0  
..                  ...             ...             ...  
374                  11              11             1.0  
375               250.5             171            86.5  
376               250.5              98           227.5  
377               450.5             337           125.0  
378               125.5             191           150.0  

[379 rows x 5 columns]

### 5. Consider only the most recent data point of the *times* dataset. Compute the number of male and female students for each country.

Per calcolare il numero di studenti maschi e femmine, consideriamo la colonna `female_male_ratio`.

In [40]:
recent['female_male_ratio']

2003    48 : 52
2056    32 : 68
1908    54 : 46
2105    48 : 52
2404    71 : 29
         ...   
2601    28 : 72
2151        NaN
2152    59 : 41
2602    43 : 57
2104    41 : 59
Name: female_male_ratio, Length: 818, dtype: object

La colonna `female_male_ratio` è costruita in modo da avere la percentuale di studenti femmine a sinistra del `:` e la percentuale di studenti maschi a destra del `:`. 

Decidiamo di costruire due funzioni (`female_ratio` e `male_ratio`) che ci permettano di ottenere, rispettivamente, la percentuale di studenti femmine e la percentuale di studenti maschi. Se la funzione trova un NaN, viene restituito un NaN. 

Successivamente, per calcolare il numero di studenti femmine e il numero di studenti maschi, basta dividere le rispettive percentuali per 100 e moltiplicarle per il numero totale di studenti dell'università corrispondente. 

In [41]:
recent['female_male_ratio'] = recent['female_male_ratio'].astype(str)

In [42]:
def female_ratio(ratio):
    if ratio == 'nan':
        return np.nan
    else:
        return ratio.split(' : ')[0] # primo elemento della lista 

In [43]:
recent['female_ratio'] = pd.to_numeric(recent.apply(lambda row: female_ratio(row.female_male_ratio), axis = 1))
recent['female_number'] = round((recent['female_ratio']/100)*recent['num_students'])

In [44]:
def male_ratio(ratio):
    if ratio == 'nan':
        return np.nan
    else:
        return ratio.split(' : ')[1] # secondo elemento della lista 

In [45]:
recent['male_ratio'] = pd.to_numeric(recent.apply(lambda row: male_ratio(row.female_male_ratio), axis = 1))
recent['male_number'] = round((recent['male_ratio']/100)*recent['num_students'])

Ora raggruppiamo per paese e vediamo il numero di studenti maschi e femmine:

In [46]:
recent.groupby('country')[['country', 'female_number', 'male_number']].sum()

female_number  male_number
country                                             
Argentina                       67191.0      41182.0
Australia                      391736.0     321640.0
Austria                         68364.0      66113.0
Bangladesh                      21323.0      41393.0
Belarus                         20219.0       9084.0
...                                 ...          ...
Uganda                          18670.0      18670.0
Ukraine                         17846.0      19250.0
United Arab Emirates             9516.0       4931.0
United Kingdom                 711815.0     613028.0
United States of America      1601453.0    1556324.0

[70 rows x 2 columns]

### 6. Find the universities where the ratio between female and male is below the average ratio (computed over all universities)

In [47]:
recent[['university_name', 'female_ratio', 'male_ratio']]

university_name  female_ratio  male_ratio
2003            aalborg university          48.0        52.0
2056              aalto university          32.0        68.0
1908             aarhus university          54.0        46.0
2105        aberystwyth university          48.0        52.0
2404    adam mickiewicz university          71.0        29.0
...                            ...           ...         ...
2601  yokohama national university          28.0        72.0
2151             yonsei university           NaN         NaN
2152               york university          59.0        41.0
2602            yuan ze university          43.0        57.0
2104           zhejiang university          41.0        59.0

[818 rows x 3 columns]

Dobbiamo trovare le università per cui il rapporto tra donne e uomini è inferiore al rapporto medio (calcolato su tutte le università). Considerare l'effettivo rapporto tra `female_ratio` e `male_ratio` è rischioso perché potremmo avere valori zero sulla colonna `male_ratio`. Infatti:

In [48]:
recent[recent['male_ratio'] == 0]

world_rank         university_name      country  teaching  international  \
2227      450.5  ewha womans university  South Korea      23.9           34.3   

      research  citations  income  total_score  num_students  \
2227      18.3       50.4    57.6          NaN       17625.0   

      student_staff_ratio international_students female_male_ratio  year  \
2227                 14.4                     6%           100 : 0  2016   

      female_ratio  female_number  male_ratio  male_number  
2227         100.0        17625.0         0.0          0.0

In [49]:
recent['ratio_femalemale'] = recent['female_ratio']/recent['male_ratio']
recent.loc[2227] # inf

world_rank                                 450.5
university_name           ewha womans university
country                              South Korea
teaching                                    23.9
international                               34.3
research                                    18.3
citations                                   50.4
income                                      57.6
total_score                                  NaN
num_students                               17625
student_staff_ratio                         14.4
international_students                        6%
female_male_ratio                        100 : 0
year                                        2016
female_ratio                                 100
female_number                              17625
male_ratio                                     0
male_number                                    0
ratio_femalemale                             inf
Name: 2227, dtype: object

Una soluzione ragionevole potrebbe essere l'eliminazione della riga nel calcolo del rapporto medio:

In [50]:
average_ratio = recent['ratio_femalemale'].drop(2227).mean()
average_ratio

1.0811020920897247

In [51]:
ratio_below_avgratio = recent[recent['ratio_femalemale'] < average_ratio]
ratio_below_avgratio[['university_name', 'country', 'ratio_femalemale']]

university_name                   country  ratio_femalemale
2003            aalborg university                   Denmark          0.923077
2056              aalto university                   Finland          0.470588
2105        aberystwyth university            United Kingdom          0.923077
2406               ajou university               South Korea          0.492537
2408         alexandria university                     Egypt          0.851852
...                            ...                       ...               ...
1814               yale university  United States of America          1.000000
2598           yeungnam university               South Korea          0.923077
2601  yokohama national university                     Japan          0.388889
2602            yuan ze university                    Taiwan          0.754386
2104           zhejiang university                     China          0.694915

[346 rows x 3 columns]

### 7. For each country, compute the fraction of the students in the country that are in one of the universities computed in the previous point (that is, the denominator of the ratio is the total number of students over all universities in the country).

Per svolgere questo punto, dobbiamo innanzitutto considerare, per ogni paese, il numero di studenti che frequentano una delle università in cui il rapporto tra femmine e maschi è inferiore al rapporto medio. Quindi consideriamo il DataFrame `ratio_below_avgratio`, raggruppiamo per paese e facciamo la somma. 

In [52]:
num_students_below = ratio_below_avgratio.groupby('country', as_index = False)['num_students'].sum()
num_students_below.rename(columns = {'num_students':'num_students_below'}, inplace = True)
num_students_below

country  num_students_below
0                  Australia            146235.0
1                    Austria             61033.0
2                 Bangladesh             62716.0
3                     Brazil            278774.0
4                     Canada             86779.0
5                      Chile             90951.0
6                      China            885293.0
7                   Colombia             54645.0
8             Czech Republic             38685.0
9                    Denmark             45705.0
10                     Egypt            359372.0
11                   Estonia             14061.0
12                   Finland             39788.0
13                    France            156153.0
14                   Germany            435244.0
15                     Ghana             37274.0
16                    Greece             57198.0
17                 Hong Kong             32079.0
18                   Hungary             22447.0
19                     India            155370.0
20                 Indonesia             43565.0
21                      Iran            107280.0
22                    Israel             30550.0
23                     Italy            106471.0
24                     Japan            447415.0
25                     Kenya             72676.0
26                   Lebanon              7695.0
27                Luxembourg              5144.0
28                  Malaysia             16270.0
29                    Mexico            151731.0
30                   Morocco             56060.0
31               Netherlands             58883.0
32               New Zealand             12187.0
33                   Nigeria             26247.0
34                    Norway             17381.0
35                      Oman              7089.0
36                  Pakistan              9808.0
37                    Poland             59789.0
38                  Portugal             89285.0
39       Republic of Ireland             34597.0
40        Russian Federation            119356.0
41                 Singapore             56620.0
42                  Slovakia             16400.0
43               South Korea            276058.0
44                     Spain            136088.0
45                    Sweden             20667.0
46               Switzerland             34692.0
47                    Taiwan            209949.0
48                  Thailand             66058.0
49                    Turkey            216017.0
50                    Uganda             37340.0
51                   Ukraine             22686.0
52      United Arab Emirates              5226.0
53            United Kingdom            259977.0
54  United States of America           1726340.0

Una volta calcolata la quantità `num_students_below`, che verrà considerata al numeratore, dobbiamo calcolare il denominatore. In particolare, quest'ultimo è pari al numero totale di studenti di tutte le università del paese. 

In [53]:
num_students_tot = recent.groupby('country', as_index = False)['num_students'].sum()
num_students_tot.rename(columns = {'num_students':'num_students_tot'}, inplace = True)
num_students_tot

country  num_students_tot
0                  Argentina          108373.0
1                  Australia          743627.0
2                    Austria          134477.0
3                 Bangladesh           62716.0
4                    Belarus           29303.0
..                       ...               ...
65                    Uganda           37340.0
66                   Ukraine           37096.0
67      United Arab Emirates           14447.0
68            United Kingdom         1324842.0
69  United States of America         3280303.0

[70 rows x 2 columns]

Ottenuta anche quest'ultima quantità, possiamo procedere con una *left join* dei due dataset in modo che vengano considerate tutte le righe di `num_students_tot` e, laddove non avessimo valori in `num_students_below`, vengano creati valori nulli. 

In [54]:
num_students_merged = pd.merge(num_students_tot, num_students_below, on = 'country', how = 'left')
num_students_merged

country  num_students_tot  num_students_below
0                  Argentina          108373.0                 NaN
1                  Australia          743627.0            146235.0
2                    Austria          134477.0             61033.0
3                 Bangladesh           62716.0             62716.0
4                    Belarus           29303.0                 NaN
..                       ...               ...                 ...
65                    Uganda           37340.0             37340.0
66                   Ukraine           37096.0             22686.0
67      United Arab Emirates           14447.0              5226.0
68            United Kingdom         1324842.0            259977.0
69  United States of America         3280303.0           1726340.0

[70 rows x 3 columns]

Sostituiamo i valori nulli con 0, in modo da avere delle frazioni anche in quei casi:

In [55]:
num_students_merged['num_students_below'] = num_students_merged['num_students_below'].fillna(0)
num_students_merged

country  num_students_tot  num_students_below
0                  Argentina          108373.0                 0.0
1                  Australia          743627.0            146235.0
2                    Austria          134477.0             61033.0
3                 Bangladesh           62716.0             62716.0
4                    Belarus           29303.0                 0.0
..                       ...               ...                 ...
65                    Uganda           37340.0             37340.0
66                   Ukraine           37096.0             22686.0
67      United Arab Emirates           14447.0              5226.0
68            United Kingdom         1324842.0            259977.0
69  United States of America         3280303.0           1726340.0

[70 rows x 3 columns]

Possiamo ora procedere calcolando la frazione richiesta (trasformandola in percentuale):

In [56]:
num_students_merged['fraction_of_students'] = (num_students_merged['num_students_below']/num_students_merged['num_students_tot'])*100
num_students_merged.sort_values(by = 'fraction_of_students', ascending = False)

country  num_students_tot  num_students_below  fraction_of_students
35  Luxembourg            5144.0              5144.0                 100.0
31       Kenya           72676.0             72676.0                 100.0
3   Bangladesh           62716.0             62716.0                 100.0
65      Uganda           37340.0             37340.0                 100.0
33     Lebanon            7695.0              7695.0                 100.0
..         ...               ...                 ...                   ...
36       Macau            8521.0                 0.0                   0.0
34   Lithuania           19019.0                 0.0                   0.0
32      Latvia           15100.0                 0.0                   0.0
30      Jordan           60263.0                 0.0                   0.0
0    Argentina          108373.0                 0.0                   0.0

[70 rows x 4 columns]

### 8. Read the file `educational_attainment_supplementary_data.csv`, discarding any row with missing `country_name` or `series_name`

Per eliminare ogni riga con anche solo un valore mancante, utilizziamo `dropna` ed inseriamo in `subset` le colonne che vogliamo tenere in considerazione. 

In [57]:
attainment = pd.read_csv("educational_attainment_supplementary_data.csv")
attainment.dropna(subset = ['country_name', 'series_name'], how = 'any', inplace = True)
attainment

country_name                                        series_name  1985  \
0      Afghanistan  Barro-Lee: Average years of primary schooling,...  0.33   
1      Afghanistan  Barro-Lee: Average years of primary schooling,...  1.03   
2      Afghanistan  Barro-Lee: Average years of primary schooling,...  0.83   
3      Afghanistan  Barro-Lee: Average years of primary schooling,...  2.34   
4      Afghanistan  Barro-Lee: Average years of primary schooling,...  0.54   
...            ...                                                ...   ...   
79045     Zimbabwe  UIS: Percentage of population age 25+ with som...   NaN   
79046     Zimbabwe  UIS: Percentage of population age 25+ with som...   NaN   
79047     Zimbabwe  UIS: Percentage of population age 25+ with unk...   NaN   
79048     Zimbabwe  UIS: Percentage of population age 25+ with unk...   NaN   
79049     Zimbabwe  UIS: Percentage of population age 25+ with unk...   NaN   

       1986  1987  1990  1991  1992  1993  1995  ...  2005  2006  2007  2008  \
0       NaN   NaN  0.44   NaN   NaN   NaN  0.57  ...  0.86   NaN   NaN   NaN   
1       NaN   NaN  1.26   NaN   NaN   NaN  1.54  ...  2.18   NaN   NaN   NaN   
2       NaN   NaN  0.95   NaN   NaN   NaN  1.26  ...  1.01   NaN   NaN   NaN   
3       NaN   NaN  2.22   NaN   NaN   NaN  2.37  ...  2.26   NaN   NaN   NaN   
4       NaN   NaN  0.92   NaN   NaN   NaN  0.94  ...  2.00   NaN   NaN   NaN   
...     ...   ...   ...   ...   ...   ...   ...  ...   ...   ...   ...   ...   
79045   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   NaN   NaN   NaN   
79046   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   NaN   NaN   NaN   
79047   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   NaN   NaN   NaN   
79048   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   NaN   NaN   NaN   
79049   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   NaN   NaN   NaN   

       2009  2010  2011      2012  2013  2015  
0       NaN  1.27   NaN       NaN   NaN   NaN  
1       NaN  2.64   NaN       NaN   NaN   NaN  
2       NaN  2.45   NaN       NaN   NaN   NaN  
3       NaN  3.55   NaN       NaN   NaN   NaN  
4       NaN  1.29   NaN       NaN   NaN   NaN  
...     ...   ...   ...       ...   ...   ...  
79045   NaN   NaN   NaN  13.84323   NaN   NaN  
79046   NaN   NaN   NaN  16.68491   NaN   NaN  
79047   NaN   NaN   NaN  11.99412   NaN   NaN  
79048   NaN   NaN   NaN   5.77150   NaN   NaN  
79049   NaN   NaN   NaN   9.14588   NaN   NaN  

[79050 rows x 29 columns]

### 9. From `attainment` build a dataframe with the same data, but with 4 columns: `country_name`, `series_name`, `year`, `value`

Consideriamo la funzione `pd.melt()` di pandas. Essa permette di trasformare il DataFrame `attainment` in un formato in cui:

* vengono mantenute le colonne `country_name` e `series_name` come nel DataFrame originale;
* i nomi delle colonne relative agli anni diventano i valori di un'unica colonna chiamata `year`;
* i valori delle colonne relative agli anni diventano i valori di un'unica colonna chiamata `value`.

In [58]:
new_attainment = attainment.melt(id_vars = ['country_name', 'series_name'], var_name = 'year', value_name = 'value')
new_attainment

country_name                                        series_name  year  \
0        Afghanistan  Barro-Lee: Average years of primary schooling,...  1985   
1        Afghanistan  Barro-Lee: Average years of primary schooling,...  1985   
2        Afghanistan  Barro-Lee: Average years of primary schooling,...  1985   
3        Afghanistan  Barro-Lee: Average years of primary schooling,...  1985   
4        Afghanistan  Barro-Lee: Average years of primary schooling,...  1985   
...              ...                                                ...   ...   
2134345     Zimbabwe  UIS: Percentage of population age 25+ with som...  2015   
2134346     Zimbabwe  UIS: Percentage of population age 25+ with som...  2015   
2134347     Zimbabwe  UIS: Percentage of population age 25+ with unk...  2015   
2134348     Zimbabwe  UIS: Percentage of population age 25+ with unk...  2015   
2134349     Zimbabwe  UIS: Percentage of population age 25+ with unk...  2015   

         value  
0         0.33  
1         1.03  
2         0.83  
3         2.34  
4         0.54  
...        ...  
2134345    NaN  
2134346    NaN  
2134347    NaN  
2134348    NaN  
2134349    NaN  

[2134350 rows x 4 columns]

### 10. For each university, find the number of rankings in which they appear (it suffices to appear in one year for each ranking).

In questo caso dobbiamo considerare tutti i dataset iniziali e, per ogni università, valutare in quante classifiche appare. 

In [59]:
times_universities = times.groupby('university_name').count()['world_rank']
times_universities = pd.DataFrame(times_universities)
times_universities.rename(columns = {'world_rank':'count_times'}, inplace=True)
times_universities

count_times
university_name                          
aalborg university                      5
aalto university                        5
aarhus university                       6
aberystwyth university                  5
adam mickiewicz university              1
...                                   ...
yokohama national university            1
yonsei university                       6
york university                         5
yuan ze university                      2
zhejiang university                     6

[818 rows x 1 columns]

In [60]:
shanghai_universities = shanghai.groupby('university_name').count()['world_rank']
shanghai_universities = pd.DataFrame(shanghai_universities)
shanghai_universities.rename(columns = {'world_rank':'count_shanghai'},inplace=True)
shanghai_universities

count_shanghai
university_name                                     
aalborg university                                 2
aalto university                                   7
aarhus university                                 11
aix marseille university                           4
aristotle university of thessaloniki               9
...                                              ...
yamaguchi university                               7
yeshiva university                                 9
yonsei university                                  9
york university                                    7
zhejiang university                               10

[650 rows x 1 columns]

In [61]:
cwur_universities = cwur.groupby('university_name').count()['world_rank']
cwur_universities = pd.DataFrame(cwur_universities)
cwur_universities.rename(columns = {'world_rank':'count_cwur'}, inplace = True)
cwur_universities

count_cwur
university_name                                
aalborg university                            2
aalto university                              2
aarhus university                             2
aberystwyth university                        2
adam mickiewicz university in pozna           2
...                                         ...
yuan ze university                            2
zhejiang normal university                    1
zhejiang university                           2
zhejiang university of technology             2
zhengzhou university                          2

[1022 rows x 1 columns]

Ora dobbiamo unire i tre dataframe.

In [62]:
conteggi = pd.merge(times_universities, shanghai_universities, left_index = True, right_index = True, how = 'outer')
conteggi = pd.merge(conteggi, cwur_universities, left_index = True, right_index = True, how = 'outer')
conteggi

count_times  count_shanghai  count_cwur
university_name                                                           
aalborg university                         5.0             2.0         2.0
aalto university                           5.0             7.0         2.0
aarhus university                          6.0            11.0         2.0
aberystwyth university                     5.0             NaN         2.0
adam mickiewicz university                 1.0             NaN         NaN
...                                        ...             ...         ...
yuan ze university                         2.0             NaN         2.0
zhejiang normal university                 NaN             NaN         1.0
zhejiang university                        6.0            10.0         2.0
zhejiang university of technology          NaN             NaN         2.0
zhengzhou university                       NaN             NaN         2.0

[1400 rows x 3 columns]

Sulla base del DataFrame conteggi, dobbiamo calcolare il numero dei rankings in cui appare l'università. Per farlo, dobbiamo contare per ogni riga i valori non nulli. 

In [63]:
conteggi['occurrence'] = conteggi.notnull().sum(axis=1)
conteggi

count_times  count_shanghai  count_cwur  \
university_name                                                              
aalborg university                         5.0             2.0         2.0   
aalto university                           5.0             7.0         2.0   
aarhus university                          6.0            11.0         2.0   
aberystwyth university                     5.0             NaN         2.0   
adam mickiewicz university                 1.0             NaN         NaN   
...                                        ...             ...         ...   
yuan ze university                         2.0             NaN         2.0   
zhejiang normal university                 NaN             NaN         1.0   
zhejiang university                        6.0            10.0         2.0   
zhejiang university of technology          NaN             NaN         2.0   
zhengzhou university                       NaN             NaN         2.0   

                                   occurrence  
university_name                                
aalborg university                          3  
aalto university                            3  
aarhus university                           3  
aberystwyth university                      2  
adam mickiewicz university                  1  
...                                       ...  
yuan ze university                          2  
zhejiang normal university                  1  
zhejiang university                         3  
zhejiang university of technology           1  
zhengzhou university                        1  

[1400 rows x 4 columns]

### 11. In the times ranking, compute the number of times each university appears

In [64]:
number_of_times = pd.DataFrame(times.groupby('university_name').size())
number_of_times = number_of_times.reset_index()
number_of_times.rename(columns = {0: 'number'}, inplace = True)
number_of_times

university_name  number
0              aalborg university       5
1                aalto university       5
2               aarhus university       6
3          aberystwyth university       5
4      adam mickiewicz university       1
..                            ...     ...
813  yokohama national university       1
814             yonsei university       6
815               york university       5
816            yuan ze university       2
817           zhejiang university       6

[818 rows x 2 columns]

### 12. Find the universities that appear at most twice in the times ranking.

Consideriamo il DataFrame precedente `number_of_times`. 

In [65]:
number_of_times[number_of_times['number']<=2]

university_name  number
4                  adam mickiewicz university       1
5    agh university of science and technology       1
6                    aix marseille university       1
7                             ajou university       1
9              alexandru ioan cuza university       1
..                                        ...     ...
808               y ld z technical university       1
811                       yeungnam university       1
812                  yokohama city university       1
813              yokohama national university       1
816                        yuan ze university       2

[408 rows x 2 columns]

### 13. The universities that, in any year, have the same position in all three rankings (they must have the same position in a year).

Per svolgere questo punto, decidiamo di costruire un dataframe che contenga: 

* una colonna contenente i nomi delle università;
* una colonna contenente l'anno;
* tre colonne contenenti, rispettivamente, i tre rankings.

Si noti che il dataframe tiene in considerazione soltanto le università che sono state classificate in tutti e tre i dataset in un certo anno. Ad esempio, se un'università è stata valutata nel 2012 soltanto da `times` e `shanghai`, ma non da `cwur`, allora la riga inerente a quella università e a quell'anno non viene riportata. 

In [66]:
all_rankings_year = pd.merge(times[['university_name', 'year', 'world_rank']], shanghai[['university_name', 'year', 'world_rank']], on = ['university_name','year'])
all_rankings_year.rename(columns = {'world_rank_x' : 'rank_times', 'world_rank_y':'rank_shanghai'}, inplace = True)
all_rankings_year

university_name  year rank_times rank_shanghai
0                    harvard university  2011          1             1
1    california institute of technology  2011          2             6
2                   stanford university  2011          4             2
3                  princeton university  2011          5             7
4               university of cambridge  2011          6             5
..                                  ...   ...        ...           ...
859                university of lisbon  2015    351-400       201-300
860                 university of turku  2015    351-400       301-400
861               university of vermont  2015    351-400       301-400
862         washington state university  2015    351-400       301-400
863                    wuhan university  2015    351-400       301-400

[864 rows x 4 columns]

In [67]:
all_rankings_year = pd.merge(all_rankings_year, cwur[['university_name', 'year', 'world_rank']], on = ['university_name','year'])
all_rankings_year.rename(columns = {'world_rank':'rank_cwur'}, inplace = True)
all_rankings_year

university_name  year rank_times rank_shanghai  \
0    california institute of technology  2012          1             6   
1                    harvard university  2012          2             1   
2                   stanford university  2012          2             2   
3                  university of oxford  2012          4            10   
4                  princeton university  2012          5             7   
..                                  ...   ...        ...           ...   
584                  university of graz  2015    351-400       401-500   
585                university of lisbon  2015    351-400       201-300   
586                 university of turku  2015    351-400       301-400   
587               university of vermont  2015    351-400       301-400   
588                    wuhan university  2015    351-400       301-400   

     rank_cwur  
0            5  
1            1  
2            3  
3            7  
4            6  
..         ...  
584        584  
585        257  
586        383  
587        324  
588        415  

[589 rows x 5 columns]

Applichiamo la funzione definita al punto 4 alle colonne `rank_times`, `rank_shanghai` e `rank_cwur` al fine di aggiustare i rankings. 

In [68]:
def adjust_rankings(x):
    if '=' in x:
        return x.replace('=','')
    else:
        if '-' in x:
            values = x.split('-')
            return (int(values[0])+int(values[1]))/2
        else:
            return x

all_rankings_year['rank_times'] = all_rankings_year['rank_times'].astype(str)
all_rankings_year['rank_shanghai'] = all_rankings_year['rank_shanghai'].astype(str)
all_rankings_year['rank_cwur'] = all_rankings_year['rank_cwur'].astype(str)

all_rankings_year['rank_times'] = all_rankings_year['rank_times'].apply(adjust_rankings)
all_rankings_year['rank_shanghai'] = all_rankings_year['rank_shanghai'].apply(adjust_rankings)
all_rankings_year['rank_cwur'] = all_rankings_year['rank_cwur'].apply(adjust_rankings)

all_rankings_year

university_name  year rank_times rank_shanghai  \
0    california institute of technology  2012          1             6   
1                    harvard university  2012          2             1   
2                   stanford university  2012          2             2   
3                  university of oxford  2012          4            10   
4                  princeton university  2012          5             7   
..                                  ...   ...        ...           ...   
584                  university of graz  2015      375.5         450.5   
585                university of lisbon  2015      375.5         250.5   
586                 university of turku  2015      375.5         350.5   
587               university of vermont  2015      375.5         350.5   
588                    wuhan university  2015      375.5         350.5   

    rank_cwur  
0           5  
1           1  
2           3  
3           7  
4           6  
..        ...  
584       584  
585       257  
586       383  
587       324  
588       415  

[589 rows x 5 columns]

Ora dobbiamo stampare soltanto le righe corrispondenti alle università che, nello stesso anno, hanno lo stesso ranking nelle tre classifiche. 

In [69]:
def equal_rankings(a,b,c):
    if a == b == c:
        return True
    else:
        return False
    
all_rankings_year[all_rankings_year.apply(lambda row: equal_rankings(row.rank_times, row.rank_shanghai, row.rank_cwur), axis = 1)]

university_name  year rank_times rank_shanghai rank_cwur
64  stanford university  2013          2             2         2

Stanford University è l'unica università che presenta i tre rankings uguali nello stesso anno. Tuttavia, per lo svolgimento del progetto, la nostra idea è stata quella di considerare il ranking medio nel caso in cui fosse stato presente un range. Pertanto, senza considerare questa operazione, sarebbe ragionevole pensare che esistano altre università per le quali il ranking nello stesso anno sia uguale in tutte e tre le classifiche. 